src/LEPTON/fix_epot_lepton.cpp

```diff
@@ -0,0 +1,324 @@
+    ilevel_respa = (dynamic_cast<Respa *>(update->integrate))->nlevels - 1;
+    if (respa_level >= 0) ilevel_respa = MIN(respa_level, ilevel_respa);
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixEpotLepton::setup(int vflag)
+{
+  if (utils::strmatch(update->integrate_style, "^respa")) {
+    auto respa = dynamic_cast<Respa *>(update->integrate);
+    respa->copy_flevel_f(ilevel_respa);
+    post_force_respa(vflag, ilevel_respa, 0);
+    respa->copy_f_flevel(ilevel_respa);
+  } else {
+    post_force(vflag);
+  }
+}
+
+/* ---------------------------------------------------------------------- */
+
+void FixEpotLepton::min_setup(int vflag)
+{
+  post_force(vflag);
+}
+
+/* ----------------------------------------------------------------------
+  Apply F = qE,
+        F = (mu . D) E,
+        T = mu x E
+------------------------------------------------------------------------- */
+
+void FixEpotLepton::post_force(int vflag)
+{
+  double **f = atom->f;
+  double **x = atom->x;
+  double *q = nullptr;
+  double **mu = nullptr, **t = nullptr;
+  int *mask = atom->mask;
+  imageint *image = atom->image;
+  int nlocal = atom->nlocal;
+
+  const double qe2f = force->qe2f;
+  const double qqr2e = force->qqr2e;
+
+  auto parsed = Lepton::Parser::parse(LeptonUtils::substitute(expr, lmp)).optimize();
+  Lepton::CompiledExpression phi, dphi_x, dphi_y, dphi_z, dphi_xx, dphi_xy, dphi_xz, dphi_yy, dphi_yz, dphi_zz; 
+
+  dphi_x = parsed.differentiate("x").createCompiledExpression();
+  dphi_y = parsed.differentiate("y").createCompiledExpression();
+  dphi_z = parsed.differentiate("z").createCompiledExpression();
+
+  std::vector<Lepton::CompiledExpression*> cexprs = {&dphi_x, &dphi_y, &dphi_z};
+
+  if (atom->q_flag) {
+    q = atom->q;
+    phi = parsed.createCompiledExpression();
+    cexprs.push_back(&phi);
+  }
+  if (atom->mu_flag) {
+    mu = atom->mu; t = atom->torque;
+    dphi_xx = parsed.differentiate("x").differentiate("x").createCompiledExpression();
+    dphi_xy = parsed.differentiate("x").differentiate("y").createCompiledExpression();
+    dphi_xz = parsed.differentiate("x").differentiate("z").createCompiledExpression();
+    dphi_yy = parsed.differentiate("y").differentiate("y").createCompiledExpression();
+    dphi_yz = parsed.differentiate("y").differentiate("z").createCompiledExpression();
+    dphi_zz = parsed.differentiate("z").differentiate("z").createCompiledExpression();
+    cexprs.insert(cexprs.end(), {&dphi_xx, &dphi_xy, &dphi_xz, &dphi_yy, &dphi_yz, &dphi_zz}); 
+  }
+
+  // check if reference to x, y, z exist
+  const std::array<std::string, 3> variableNames = {"x", "y", "z"}; 
+  std::vector<std::array<bool, 3>> has_ref;
+  for (auto &cexpr : cexprs) {
+    has_ref.push_back({true, true, true});
+    for (size_t i = 0; i < 3; i++) {
+      try { 
+        (*cexpr).getVariableReference(variableNames[i]);
+      }
+      catch (Lepton::Exception &) {
```

Comment:
All of this try-catching feels expensive. If we can't move this out of the inner force loop (!!), I wonder if we can just add `+0*(x+y+z)*(x+y+z)` to the original expression and call it a day. 😈